465 animal observation summary

cookeac opened the pull request:
I tried to add proposed commits to #467 but didn't have permissions. 
This is a draft PR for review and comments. It addresses recommendations from the ADE meetings to address #465.

- The initial examples are (time-based) summaries of observations (e.g. the number of minutes in a day spent in an activity, or the number of steps in a day).
- The `icarStatisticsType` (used in group-based `icarStatisticsResource`) is close to what we need but we need to address time period or duration
- Address naming conventions - avoid the use of bare "type" and consider using "metric" as we use this for similar things elsewhere.
- Recognise that we don't need "DailyRuminationTime" and "24HourRuminationTime" - these are both different duration or period summaries for the same metric.

If we are on the right track, I will copy in the URL scheme that has already been defined for #467.

cookeac committed:
Merge pull request #22 from adewg/Develop

Update from Develop

cookeac committed:
Add Observation Summary Resouce

Initial commit to add icarObservationSummaryResource for animal activity and similar observations. Defines an icarObservationStatisticsType that extends icarStatisticsType, and an activity metrics list.

Resolves #465

erwinspeybroeck commented:
1) typo in the name of resources/icar**l**ObservationSummaryResource.json

2) a lot of redundancy?

for each animal the location/dateFrom/dateTo --> is it possible that this changes within the same response?

for each statistic the startDateTime and duration --> is it possible that this changes within the same response?

I think the response will always contain data for the same location, dates, periods, so 

{ location, date from, dateTo, startDateTime, duration (or you use dateFrom - dateTo, or you use startDateTime + duration)
within this an array of animals
within each animal an array of statistics

we also add an id on the main level - don't know if that is necessary

for the animal we alsi had a request to add the parity at the moment of the observation - i'm not 100% in favor, but ...

Of course a receiver will probably store this  in a "flat" way on animal level:
animal, location, date info, metric info --> so without any arrays. Don't know if we have to take that into account.

cookeac commented:
> 1. typo in the name of resources/icar**l**ObservationSummaryResource.json

Oops, will fix

> 2. a lot of redundancy?

Definitely there is redundancy in the `icarObservationSummaryResource` having location, dateFrom, and dateTo which will also be parameters to the API call. These could be removed entirely to my mind, but I am keen for other people's thoughts.

The statistics are an array of "daily" data (or whatever periods are provided), so startDateTime of a statistic will vary for every statistic for an animal. Similarly the duration may vary for each statistic, returning (for instance) "D" or the "H24" in the same API call. So there is no real redundancy for these.

cookeac committed:
Fix file name typo and clarify descriptions.

cookeac committed:
Add isIncomplete as per comments on #465

cookeac commented on enums/icarDurationType.json:
```diff
@@ -0,0 +1,14 @@
+    "24H",
+    "1W",
+    "1M"
+  ]
```
Add 96H for compatibility with ICAR recommendations

cookeac commented on resources/icarObservationSummaryResource.json:
```diff
@@ -0,0 +1,40 @@
+            "$ref": "../types/icarAnimalIdentifierType.json",
+            "description": "Unique animal scheme and identifier combination."
+          },
+          "location": {
```
Remove location - as this is in icarResource

cookeac commented on resources/icarObservationSummaryResource.json:
```diff
@@ -0,0 +1,40 @@
+            "$ref": "../types/icarLocationIdentifierType.json",
+            "description": "Location where the observations occurred."
+          },
+          "dateFrom": {
```
remove dateFrom and dateTo as they are in the API query parameters

cookeac committed:
Improvements to icarObservationSummaryResource

1. Extend icarAggregationType with "Range" and "Score".
2. Add icarObservationSummaryCollection.
3. Remove location, dateFrom, dateTo from icarObservationSummaryResource.
4. Rename icarActivityMetrics markdown to icarObservationSummaryMetrics
5. Define URL schemes in exampleURLScheme and managementURLScheme (GET only).

cookeac committed:
Add 96H to duration

NedapLivestockManagement commented on collections/icarObservationSummaryCollection.json:
```diff
@@ -0,0 +1,21 @@
+{
+  "description": "Represents a collection of animal sets. Based on icarResourceCollection to provide paging etc.",
```
should be: Represents a collection of observation summary statistics.

cookeac replied:
```suggestion
  "description": "Represents a collection of observation summary statistics. Based on icarResourceCollection to provide paging etc.",
```

cookeac committed:
Update collections/icarObservationSummaryCollection.json

Incorporate suggestion correcting description

cookeac commented:
Thanks for the feedback and adjustments. Based on our discussion on Nov 27, 2004 we are merging this.
Resolves #465
